src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -54,10 +67,199 @@ import qualified Data.ByteString.Short.Internal as SBS
+    p <- readSTRef currentPageRef
+    mlastPage <-
+      if paIsEmpty p then
+        pure Nothing
+      else do
+        newChunks <- append (mkAppend p) mindex
+        modifySTRef' indexChunksRef $ \chunkss -> newChunks : chunkss
+        pure $ Just $ pageBuilder p
+    (chunk, fchunk) <- Index.unsafeEnd mindex
+    bloom <- Bloom.freeze mbloom
+    chunkss <- readSTRef indexChunksRef
+    let chunks = concat (reverse ([chunk] : chunkss))
+        index = Index.fromChunks chunks fchunk
+    pure (mlastPage, chunk, fchunk, bloom, index)
+
+{-------------------------------------------------------------------------------
+  Adding k\/op pairs
+-------------------------------------------------------------------------------}
+
+type Add s = MRun s -> ST s (Maybe (SizedBuilder, [Index.Chunk]))
+
+-- | Add a serialised k\/op pair with an optional blob reference.
+--
+-- Use only for full values.
+add :: KOp -> Add s
+add kop@(k0, _, _) mrun@MRun{..} = do
+    Bloom.insert mbloom k0
+    p <- readSTRef currentPageRef
+    case paAddElem kop p of
+      Nothing -> do
+        (pb, cs) <- yieldAlways mrun
+        let p' = paSingleton kop
+        writeSTRef currentPageRef $! p'
+        pure $ Just (pb, cs)
+      Just p' -> do
+        writeSTRef currentPageRef $! p'
+        pure Nothing
+
+-- | Add a serialised Insert operation.
+add_insert :: RawKey -> RawValue -> Add s
+add_insert k v = add (k, Insert v, Nothing)
+
+-- | Add a serialised Insert operation with a blob reference.
+add_insertWithBlob ::
+     RawKey
+  -> RawValue
+  -> BlobRef
+  -> Add s
+add_insertWithBlob k v b = add (k, Insert v, Just b)
+
+-- | Add a serialised Mupsert operation.
+add_mupsert :: RawKey -> RawValue -> Add s
+add_mupsert k v = add (k, Mupsert v, Nothing)
+
+-- | Add a serialised Delete operation.
+add_delete :: RawKey -> Add s
+add_delete k = add (k, Delete, Nothing)
+
+-- | __For multi-page values only during run merging:__ add the first chunk of a
+-- serialised k\/op pair with an optional blob reference. The caller of this
+-- function is responsible for copying the remaining chunks to the output file.
+--
+-- Will throw an error if the result of adding the k\/op pair is not exactly the
+-- size of a disk page.
+--
+-- During a merge, raw values that span multiple pages may not be read into memory
+-- as a whole, but in chunks. If you have access to the full bytes of a raw
+-- value, use 'add' and co instead.
+addMultiPageInChunks :: KOp -> Word32 -> MRun s -> ST s (SizedBuilder, [Index.Chunk])
+addMultiPageInChunks kop nOverflow mrun@MRun{..}
+  | pageSizeNumBytes (psSingleton kop) == 4096 = do
+      prev <- yield mrun
+
+      Bloom.insert mbloom (kopKey kop)
+      cs' <- append (AppendMultiPage (kopKey kop) nOverflow) mindex
+      storeChunks mrun cs'
+      let p'  = paSingleton kop
+          pb' = pageBuilder p'
+
+      case prev of
+        Nothing       -> pure (pb', cs')
+        Just (pb, cs) -> pure (pb <> pb', cs' <> cs)
```

Comment:
`pb` might not be a full page, so if we want to append the two builders we need padding inbetween. Since we don't deal with padding elsewhere in this module, would it make sense to just return `prev` along with `pb'` and `cs'` and let the caller take care of it? (or return lists of 1-2 elements)